big-red-shuttle/GPS.swift

```diff
@@ -0,0 +1,125 @@
+import UIKit
+import SwiftyJSON
+
+@objc protocol ShuttleBusGPSDelegate: class {
+    
+    @objc optional func gps(gps: GPS, movedToCoordinate coordinate: Coordinate)
+    @objc optional func gpsBeganFetching(gps: GPS)
+    @objc optional func gpsStoppedFetching(gps: GPS)
+    @objc optional func gpsBeganLogging(gps: GPS)
+    @objc optional func gpsStoppedLogging(gps: GPS)
+    
+}
+
+@objc class GPS: NSObject {
+    
+    static let shared = GPS()
+    var observers = [ShuttleBusGPSDelegate?]()
+    var fetchUpdateFrequency: Double = 5.0
+    var logUpdateFrequency: Double = 5.0
+    var isFetchingShuttleLocation = false
+    var isLoggingShuttleLocation = false
+    
+    override init() {
+        super.init()
+    }
+    
+    func addObserver(observer: ShuttleBusGPSDelegate) {
+        weak var weakObserver = observer
+        observers.append(weakObserver)
+    }
+    
+    func removeObserver(observer: ShuttleBusGPSDelegate) {
+        weak var weakObserver = observer
+        for i in 0..<observers.count {
+            if weakObserver === observers[i] {
+                observers.remove(at: i)
+                return
+            }
+        }
+    }
+    
+    //MARK: - Logging
+    
+    func beginLoggingShuttleLocation() {
+        isLoggingShuttleLocation = true
+        notifyObserversGPSBeganLogging()
+        logShuttleLocation()
+    }
+    
+    func stopLoggingShuttleLocation() {
+        isLoggingShuttleLocation = false
+        notifyObserversGPSStoppedLogging()
+    }
+    
+    func logShuttleLocation() {
+        if !isLoggingShuttleLocation { return }
+        
+        api().logLocation(success: { (json: JSON?) in
+            Timer.scheduledTimer(timeInterval: self.fetchUpdateFrequency,
+                                 target: self,
+                                 selector: #selector(self.fetchShuttleLocation),
+                                 userInfo: nil,
+                                 repeats: false)
+        }, failure: { (json: JSON?) in
```

Comment:
I actually don't like using _ because then devs can't figure out what that variable is at a quick glance. Similar to why I leave the types in the block declarations